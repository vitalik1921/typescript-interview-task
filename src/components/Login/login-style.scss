.login-page {
  background: url("/public/assets/mountain.jpg");
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;

  .login-form {
    display: flex;
    position: relative;
    justify-content: center;
    flex-direction: column;
    background-color: white;
    padding: 50px;
    width: 350px;
    height: 400px;

    &.loading:after {
      content: "";
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      display: block;
      background-color: rgba(255, 255, 255, 0.5);
    }
    &.loading:before {
      content: "";
      position: absolute;
      display: block;
      top: 50%;
      left: 50%;
      width: 64px;
      height: 64px;
      transform: translate(-50%, -50%);
      background: url("../../icons/loader.svg");
      background-size: contain;
    }
  }
}
