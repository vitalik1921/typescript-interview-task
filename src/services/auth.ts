import { API } from "~/constants";
import getUrl from "~/utils/getUrl";

export const login = async (username: string, password: string) => {
  const url = getUrl(API.Login);

  const bodyString = JSON.stringify({
    username,
    password,
  });

  const response = await fetch(url, {
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      "Content-Length": bodyString.length.toString(),
    },
    method: "POST",
    body: bodyString,
  });
  const data = await response.json();
  const { token } = data;

  localStorage.setItem("token", token);
};

export const logout = async () => {
  const url = getUrl(API.Logout);
  await fetch(url);
  localStorage.removeItem("token");
};
