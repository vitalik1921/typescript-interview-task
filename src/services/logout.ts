import { API } from "~/constants";
import getUrl from "~/utils/getUrl";

const logout = async () => {
  const url = getUrl(API.Logout);
  await fetch(url);
  localStorage.removeItem("token");
};

export default logout;
