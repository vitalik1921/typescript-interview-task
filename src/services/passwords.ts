import { API } from "~/constants";
import getUrl from "~/utils/getUrl";

export interface IPassword {
  id: string;
  title: string;
  description: string;
  password: string;
  createdAt: string;
}

export const getPasswords = async (userId?: string): Promise<Array<IPassword>> => {
  const url = getUrl(API.Items, {
    userId,
  });

  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${localStorage.getItem("token")}`,
    },
  });

  const data = await response.json();

  return data.items;
};

export const updatePassword = async (item: IPassword): Promise<IPassword> => {
  const response = await fetch(getUrl(API.Items), {
    method: "POST",
    body: JSON.stringify(item),
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${localStorage.getItem("token")}`,
    },
  });

  return await response.json();
};
