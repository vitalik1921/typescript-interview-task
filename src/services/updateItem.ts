import { API } from "~/constants";
import getUrl from "~/utils/getUrl";
import { IItem } from "./getUserItems";

const updateItem = async (item: IItem): Promise<IItem> => {
  const response = await fetch(getUrl(API.Items), {
    method: "POST",
    body: JSON.stringify(item),
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${localStorage.getItem("token")}`,
    },
  });

  return await response.json();
};

export default updateItem;
