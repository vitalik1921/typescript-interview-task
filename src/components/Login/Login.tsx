import "./login-style.scss";

import { SyntheticEvent, useState } from "react";

import { useHistory } from "react-router-dom";
import { Routes } from "~/constants";
import { login } from "~/services/auth";

import ErrorBlock from "../ErrorBlock";

const Login = () => {
  const { push } = useHistory();
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [loading, setLoading] = useState(false);
  const [errorMessage, setErrorMessage] = useState<string>();

  const handleSubmit = async (event: SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault();
    setErrorMessage(null);
    setLoading(true);

    try {
      await login(username, password);
      push(Routes.PasswordHealth);
    } catch (error) {
      setErrorMessage("Invalid username or password");
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="login-page">
      <form
        className={`login-form ${loading ? "loading" : ""}`}
        onSubmit={handleSubmit}
      >
        <h1 className="text-center">Password Health</h1>
        <input
          value={username}
          onChange={(event) => setUsername(event.target.value)}
          placeholder="Username"
          type="text"
          className="input mt-52px"
        />
        <input
          value={password}
          onChange={(event) => setPassword(event.target.value)}
          placeholder="Password"
          type="password"
          className="input mt-24px"
        />
        <ErrorBlock error={errorMessage} />
        <button type="submit" className="button mt-24px">
          Login
        </button>
      </form>
    </div>
  );
};

export default Login;
