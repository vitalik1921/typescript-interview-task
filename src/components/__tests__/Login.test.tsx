import * as auth from "~/services/auth";

import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";

import Login from "../Login/Login";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

afterEach(cleanup);

describe("<Login /> component", () => {
  it("matches snapshot", async () => {
    const { asFragment, container } = render(<Login />);
    expect(container).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it("submits form", async () => {
    const { getByTestId } = render(<Login />);
    const loginMock = jest.spyOn(auth, "login");
    loginMock.mockResolvedValue();

    const loginForm = getByTestId("login-form");
    const usernameInput = getByTestId("username-input");
    const passwordInput = getByTestId("password-input");
    const submitButton = getByTestId("submit-button");

    fireEvent.change(usernameInput, { target: { value: "testUser" } });
    fireEvent.change(passwordInput, { target: { value: "testPass" } });
    fireEvent.click(submitButton);

    expect(loginForm.classList.contains("loading")).toBeTruthy();
    expect(loginMock).toBeCalledWith("testUser", "testPass");
    await waitFor(() => expect(mockHistoryPush).toBeCalled());
    mockHistoryPush.mockClear();
  });

  it("displays errors", async () => {
    const { getByTestId, getByText } = render(<Login />);
    const loginMock = jest.spyOn(auth, "login");
    loginMock.mockRejectedValue(false);

    const loginForm = getByTestId("login-form");
    const submitButton = getByTestId("submit-button");
    fireEvent.click(submitButton);

    expect(loginForm.classList.contains("loading")).toBeTruthy();
    await waitFor(() => expect(mockHistoryPush).not.toBeCalled());
    await waitFor(() =>
      expect(getByText("Invalid username or password")).toBeDefined()
    );
    await waitFor(() =>
      expect(loginForm.classList.contains("loading")).toBeFalsy()
    );
  });
});
