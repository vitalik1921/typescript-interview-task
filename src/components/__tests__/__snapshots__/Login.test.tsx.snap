// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="login-page"
  >
    <form
      class="login-form "
      data-testid="login-form"
    >
      <h1
        class="text-center"
      >
        Password Health
      </h1>
      <input
        class="input mt-52px"
        data-testid="username-input"
        placeholder="Username"
        type="text"
        value=""
      />
      <input
        class="input mt-24px"
        data-testid="password-input"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        class="button mt-24px"
        data-testid="submit-button"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;
